pkg/controller/framework/shared_informer.go

```diff
@@ -0,0 +1,243 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package framework
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"k8s.io/kubernetes/pkg/client/cache"
+	"k8s.io/kubernetes/pkg/runtime"
+	utilruntime "k8s.io/kubernetes/pkg/util/runtime"
+)
+
+// if you use this, there is one behavior change.  When you receive a notification, the cache will be AT LEAST as
+// current as the notification, but it MAY be more current.  So if there was a create, followed by a delete, the cache
```

Comment:
I'm not sure I understand it. Do you mean that in base Informer handler functions are called in the same thread where cache is being updated, so there is guarantee that you can access it from handler func and the state is exactly from that point?
Whereas in the SharedInformer, this is no longer the case (so there is a different thread for calling handler functions and it may potentially lag?)

I think this is fine, but we should explicitly say something like:
"You should NOT depend on the contents of the cache in handler functions, as there is no guarantee on its contents."
(or sth like that).